pages/models.py

```diff
@@ -548,64 +548,197 @@ def get_context(self, *args, **kwargs):
+        'wagtailimages.Image',
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name='+'
+    )
+    state_or_territory = models.CharField(null=True, blank=True, max_length=32)
+    twitter_url = models.URLField(null=True, blank=True)
+    volunteer_url = models.URLField(null=True, blank=True)
+    website_url = models.URLField(null=True, blank=True)
+    youtube_url = models.URLField(null=True, blank=True)
 
+    parent_page_types = ['pages.CandidateEndorsementIndexPage']
 
     content_panels = Page.content_panels + [
-            FieldPanel('body', classname="full"),
-            FieldPanel('candidate'),
-            FieldPanel('election')
-        ]
+        MultiFieldPanel(
+            [
+                FieldPanel('body', classname="full"),
+                ImageChooserPanel('photo'),
+            ],
+            heading="Candidate",
+            classname="collapsible"
+        ),
+        MultiFieldPanel(
+            [
+                FieldPanel('office'),
+                FieldPanel('state_or_territory'),
+                FieldPanel('election'),
+                FieldPanel('primary_election_date'),
+                FieldPanel('primary_election_result'),
+                FieldPanel('general_election_date'),
+                FieldPanel('general_election_result'),
+            ],
+            heading="Election",
+            classname="collapsible"
+        ),
+        MultiFieldPanel(
+            [
+                FieldPanel('donate_url'),
+                FieldPanel('volunteer_url'),
+                FieldPanel('website_url'),
+                FieldPanel('twitter_url'),
+                FieldPanel('facebook_url'),
+                FieldPanel('youtube_url'),
+                FieldPanel('instagram_url'),
+            ],
+            heading="Links",
+            classname="collapsible"
+        ),
+        MultiFieldPanel(
+            [FieldPanel('candidate')],
+            heading="Legacy fields",
+            classname="collapsible collapsed"
+        ),
+    ]
 
     promote_panels = Page.promote_panels + [
-            ImageChooserPanel('social_image')
-        ]
+        ImageChooserPanel('social_image')
+    ]
+
+    '''
+    Get election date for general or primary depending on which is relevant
+    '''
+    def _get_election_date(self):
+        # Return primary date if it is active
+        if (
+            self.general_election_result is None and
+            self.primary_election_date is not None and
+            self.primary_election_result is None
+        ):
+            return self.primary_election_date
+        # Return primary date if candidate lost primary
+        elif (
+            self.general_election_result is None and
+            self.primary_election_result == 'loss'
+        ):
+            return self.primary_election_date
+        # Return general date otherwise
+        else:
+            return self.general_election_date
```

Comment:
This can fail if `general_election_date` is not required. Either need to make `general_election_date` required or provide fallback here.

Reply:
It will return None as a fallback, I think that's fine? Since the fields are optional then this function also returns an optional value basically.

Reply:
👍